### Four main types of numbers:

1. Integer numbers (Z)  - Python Type: int
2. Rational numbers (Q) - Python Type: fractions.Fraction
3. Real numbers (R)     - Python Type: float, decimal.Decimal
4. Complex numbers (C)  - Python Type: complex

Remember: Z is subset of Q is subset of R is subset of C --> C is the biggest subset

### In Python, there are Five main types of numbers:

1. Boolean truth values - Python Type: bool
2. Integer numbers (Z)  - Python Type: int
3. Rational numbers (Q) - Python Type: fractions.Fraction
4. Real numbers (R)     - Python Type: float, decimal.Decimal
5. Complex numbers (C)  - Python Type: complex

In [1]:
# what the large magitude
a = int('10011', 2) # base 2
b = int('19', 10) # base 10

print(a, b)

19 19


### a and b are presented above 
how large can a Python int become (posivive or negative)?

decimal number 19 required 5 bits (10011) in order to store or represent it

In [2]:
# what is the largeest integer number can be presented using 8 bits?

c = int('11111111', 2) # base 2
c


255

### the largest number that we can use is 255
255 = 2**8 - 1

In [4]:
2**8 - 1

255

In [8]:
# now let check for the negative number

d = int('-1111111', 2) # base 2 ( 1 bit is reserved to represent the sign of number)
d


-127

### The largest number that we can represent using 7 bits is 127
2**7 - 1 = 127

In [9]:
2**7 - 1

127

## So, using 8 bits we can represent all the integers in the range of [-127, 127]

and 0 dese not require a sign then we can end up with the range of [-128, 127]

### [-2^7, 2^7 - 1]

### using 16 bits to store (signed) integers
2^(16-1) = 2^15 Range is [-32768 ... 32767]

### using 32 bits to store (signed) integers
2^(32-1) = 2^31 Range is [-2^31 ... 2^31 -1 ]


In [10]:
-2**31

-2147483648

In [14]:
2**31 - 1

2147483647